src/core/experimental/array.d

```diff
@@ -0,0 +1,2559 @@
+    }
+
+    private template Payload2Affix(Payload, Affix)
+    {
+        static if (is(Payload[] : void[]))
+            alias Payload2Affix = Affix;
+        else static if (is(Payload[] : shared(void)[]))
+            alias Payload2Affix = shared Affix;
+        else static if (is(Payload[] : immutable(void)[]))
+            alias Payload2Affix = shared Affix;
+        else static if (is(Payload[] : const(shared(void))[]))
+            alias Payload2Affix = shared Affix;
+        else static if (is(Payload[] : const(void)[]))
+            alias Payload2Affix = const Affix;
+        else
+            static assert(0, "Internal error for type " ~ Payload.stringof);
+    }
+
+    static auto ref prefix(T)(T[] b)
+    {
+        assert(b.ptr && (cast(size_t) b.ptr % alignment == 0));
+        return (cast(Payload2Affix!(T, size_t)*) b.ptr)[-1];
+    }
+
+    /**
+    Returns the global instance of this allocator type. The C heap allocator is
+    thread-safe, therefore all of its methods and `it` itself are `shared`.
+    */
+    static shared PrefixAllocator instance;
+}
+
+@safe pure nothrow @nogc unittest
+{
+    shared PrefixAllocator a;
+    auto b = a.allocate(42);
+    assert(b.length == 42);
+    assert(a.bytesUsed == 42);
+    () @trusted {
+        assert(a.prefix(b) == 0);
+        a.prefix(b)++;
+        assert(a.prefix(b) == 1);
+        a.deallocate(b);
+    }();
+    assert(a.bytesUsed == 0);
+}
+
+
+version(unittest)
+{
+    private alias SCAlloc = shared PrefixAllocator;
+    private alias SSCAlloc = shared PrefixAllocator;
+
+    SCAlloc _allocator;
+    SSCAlloc _sallocator;
+
+    @nogc nothrow pure @trusted
+    void[] pureAllocate(bool isShared, size_t n)
+    {
+        return (cast(void[] function(bool, size_t) @nogc nothrow pure)(&_allocate))(isShared, n);
+    }
+
+    @nogc nothrow @safe
+    void[] _allocate(bool isShared, size_t n)
+    {
+        return isShared ? _sallocator.allocate(n) : _allocator.allocate(n);
+    }
+
+    static if (hasMember!(typeof(_allocator), "expand"))
+    {
+        @nogc nothrow pure @trusted
+        bool pureExpand(bool isShared, ref void[] b, size_t delta)
+        {
+            return (cast(bool function(bool, ref void[], size_t) @nogc nothrow pure)(&_expand))(isShared, b, delta);
+        }
+
+        @nogc nothrow @safe
+        bool _expand(bool isShared, ref void[] b, size_t delta)
+        {
+            return isShared ?  _sallocator.expand(b, delta) : _allocator.expand(b, delta);
+        }
+    }
+
+    @nogc nothrow pure
+    void pureDispose(T)(bool isShared, T[] b)
+    {
+        return (cast(void function(bool, T[]) @nogc nothrow pure)(&_dispose!(T)))(isShared, b);
+    }
+
+    @nogc nothrow
+    void _dispose(T)(bool isShared, T[] b)
+    {
+        return isShared ?  _sallocator.dispose(b) : _allocator.dispose(b);
+    }
+}
+
+///
+struct Array(T)
+{
+    import core.atomic : atomicOp;
+
+    package T[] _payload;
+    package Unqual!T[] _support;
```

Comment:
Some developers have that habit, especially C++ developers. I recommend against it. It should only be used if it would otherwise conflict with a keyword or a method, but the. It should be a trailing underscore.